File: app/src/main/java/com/genonbeta/TrebleShot/task/OrganizeLocalSharingTask.java
/*
 * Copyright (C) 2020 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package com.genonbeta.TrebleShot.task;

import android.content.Context;
import android.content.Intent;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;
import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.activity.TransferDetailActivity;
import com.genonbeta.TrebleShot.activity.TransferMemberActivity;
import com.genonbeta.TrebleShot.activity.WebShareActivity;
import com.genonbeta.TrebleShot.adapter.FileListAdapter;
import com.genonbeta.TrebleShot.database.Kuick;
import com.genonbeta.TrebleShot.io.Containable;
import com.genonbeta.TrebleShot.object.Container;
import com.genonbeta.TrebleShot.object.Shareable;
import com.genonbeta.TrebleShot.object.Transfer;
import com.genonbeta.TrebleShot.object.TransferItem;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableAsyncTask;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener;
import com.genonbeta.TrebleShot.service.backgroundservice.TaskMessage;
import com.genonbeta.TrebleShot.service.backgroundservice.TaskStoppedException;
import com.genonbeta.TrebleShot.util.AppUtils;
import com.genonbeta.TrebleShot.util.FileUtils;
import com.genonbeta.TrebleShot.util.Transfers;
import com.genonbeta.android.framework.io.DocumentFile;

import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;

public class OrganizeLocalSharingTask extends AttachableAsyncTask<AttachedTaskListener>
{
    public static final String TAG = OrganizeLocalSharingTask.class.getSimpleName();

    public List<? extends Shareable> mList;
    private final boolean mFlagAddNewDevice;
    private final boolean mFlagWebShare;

    public OrganizeLocalSharingTask(List<? extends Shareable> list, boolean addNewDevice, boolean webShare)
    {
        mList = list;
        mFlagAddNewDevice = addNewDevice;
        mFlagWebShare = webShare;
    }

    @Override
    protected void onRun() throws TaskStoppedException
    {
        if (mList.size() <= 0)
            return;

        final Kuick kuick = AppUtils.getKuick(getContext());
        final SQLiteDatabase db = kuick.getWritableDatabase();
        final Transfer transfer = new Transfer(AppUtils.getUniqueNumber());
        final List<TransferItem> list = new ArrayList<>();

        progress().addToTotal(mList.size());

        for (Shareable shareable : mList) {
            throwIfStopped();
            progress().addToCurrent(1);
            publishStatus();

            Containable containable = shareable instanceof Container ? ((Container) shareable).expand() : null;

            if (shareable instanceof FileListAdapter.FileHolder) {
                DocumentFile file = ((FileListAdapter.FileHolder) shareable).file;
                if (file != null) {
                    if (file.isDirectory())
                        Transfers.createFolderStructure(list, transfer.id, file, shareable.fileName, this,
                                progressListener());
                    else
                        list.add(TransferItem.from(file, transfer.id, null));
                }
            } else
                list.add(TransferItem.from(shareable, transfer.id, containable == null ? null : shareable.friendlyName));

            if (containable != null)
                for (Uri uri : containable.children)
                    try {
                        list.add(TransferItem.from(FileUtils.fromUri(getContext(), uri), transfer.id,
                                shareable.friendlyName));
                    } catch (FileNotFoundException e) {
                        e.printStackTrace();
                    }
        }

        if (list.size() <= 0) {
            post(TaskMessage.newInstance()
                    .setTitle(getContext(), R.string.text_error)
                    .setMessage(getContext(), R.string.text_errorNoFileSelected));
            Log.d(TAG, "onRun: No content is located with uri data");
            return;
        }

        addCloser((userAction -> kuick.remove(db, transfer, null, null)));
        kuick.insert(db, list, transfer, progressListener());

        if (mFlagWebShare) {
            transfer.isServedOnWeb = true;

            new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(getContext(),
                    R.string.text_transferSharedOnBrowser, Toast.LENGTH_SHORT).show());
        }

        kuick.insert(db, transfer, null, progressListener());
        TransferDetailActivity.startInstance(getContext(), transfer);

        if (mFlagWebShare)
            getContext().startActivity(new Intent(getContext(), WebShareActivity.class).addFlags(
                    Intent.FLAG_ACTIVITY_NEW_TASK));
        else
            TransferMemberActivity.startInstance(getContext(), transfer, mFlagAddNewDevice);

        kuick.broadcast();
    }

    @Override
    public String getName(Context context)
    {
        return context.getString(R.string.mesg_organizingFiles);
    }

}

File: app/src/main/java/com/genonbeta/TrebleShot/task/OrganizeLocalSharingTask.kt
/*
 * Copyright (C) 2020 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.task

import android.content.Context
import android.content.Intent
import android.database.sqlite.SQLiteDatabase
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import com.genonbeta.TrebleShot.R
import com.genonbeta.TrebleShot.activity.TransferMemberActivity
import com.genonbeta.TrebleShot.activityimport.WebShareActivity
import com.genonbeta.TrebleShot.adapter.FileListAdapter
import com.genonbeta.TrebleShot.dataobject.Container
import com.genonbeta.TrebleShot.dataobject.Shareable
import com.genonbeta.TrebleShot.dataobject.Transfer
import com.genonbeta.TrebleShot.dataobject.TransferItem
import com.genonbeta.TrebleShot.dataobject.TransferItem.Companion.from
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableAsyncTask
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener
import com.genonbeta.TrebleShot.service.backgroundservice.TaskMessage
import com.genonbeta.TrebleShot.service.backgroundservice.TaskStoppedException
import com.genonbeta.TrebleShot.util.AppUtils
import com.genonbeta.TrebleShot.util.Transfers
import com.genonbeta.android.framework.util.Files
import com.genonbeta.android.framework.util.Stoppable
import java.io.FileNotFoundException
import java.util.*

class OrganizeLocalSharingTask(
    var list: List<Shareable>,
    private val flagAddNewDevice: Boolean,
    private val flagWebShare: Boolean,
) : AttachableAsyncTask<AttachedTaskListener>() {
    @Throws(TaskStoppedException::class)
    override fun onRun() {
        if (list.isEmpty())
            return

        val db: SQLiteDatabase = kuick.writableDatabase
        val transfer = Transfer(AppUtils.uniqueNumber.toLong())
        val list: MutableList<TransferItem> = ArrayList()

        progress.increaseTotalBy(this.list.size)

        for (shareable in this.list) {
            throwIfStopped()
            ongoingContent = shareable.fileName
            progress.increaseBy(1)

            val containable = if (shareable is Container) (shareable as Container).expand() else null

            if (shareable is FileListAdapter.FileHolder) {
                if (shareable.file.isDirectory()) Transfers.createFolderStructure(
                    list,
                    transfer.id,
                    shareable.file,
                    shareable.fileName,
                    this
                ) else
                    list.add(from(shareable.file, transfer.id, null))
            } else
                list.add(from(shareable, transfer.id, if (containable == null) null else shareable.friendlyName))
            if (containable != null) {
                progress.increaseTotalBy(containable.children.size)

                for (uri in containable.children) {
                    progress.increaseBy(1)
                    try {
                        list.add(
                            from(
                                Files.fromUri(context, uri),
                                transfer.id,
                                shareable.friendlyName
                            )
                        )
                    } catch (e: FileNotFoundException) {
                        e.printStackTrace()
                    }
                }
            }
        }
        if (list.size <= 0) {
            post(
                TaskMessage.newInstance(
                    context.getString(R.string.text_error),
                    context.getString(R.string.text_errorNoFileSelected)
                )
            )
            Log.d(TAG, "onRun: No content is located")
            return
        }
        addCloser(object : Stoppable.Closer {
            override fun onClose(userAction: Boolean) {
                kuick.remove(db, transfer, null, null)
            }
        })
        kuick.insert(db, list, transfer, progress)
        if (flagWebShare) {
            transfer.isServedOnWeb = true
            Handler(Looper.getMainLooper()).post {
                Toast.makeText(
                    context,
                    R.string.text_transferSharedOnBrowser, Toast.LENGTH_SHORT
                ).show()
            }
        }
        kuick.insert(db, transfer, null, progress)
        if (flagWebShare) context.startActivity(
            Intent(context, WebShareActivity::class.java).addFlags(
                Intent.FLAG_ACTIVITY_NEW_TASK
            )
        ) else
            TransferMemberActivity.startInstance(context, transfer, flagAddNewDevice)
        kuick.broadcast()
    }

    override fun getName(context: Context): String {
        return context.getString(R.string.mesg_organizingFiles)
    }

    companion object {
        val TAG = OrganizeLocalSharingTask::class.java.simpleName
    }
}